File: 448-find-all-numbers-disappeared-in-an-array/find-all-numbers-disappeared-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDisappearedNumbers = function(nums) {
    let set = new Set(nums);
    let n = nums.length;
    let array = [];
    
    for (let i = 1; i <= n; i++) {
        if (!set.has(i)) {
            array.push(i)
        }
    }
    return array
};

File: 448-find-all-numbers-disappeared-in-an-array/find-all-numbers-disappeared-in-an-array.java
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        ArrayList<Integer> res = new ArrayList<>();
        int i = 0;
        while (i < nums.length) {
            int correct = nums[i] - 1;
            if (nums[i] != nums[correct]) {
                swap(nums, i, correct);
            } else {
                i++;
            }
        }

        for (int idx = 0; idx < nums.length; idx++) {
            if (nums[idx] != idx + 1) {
                res.add(idx + 1);
            }
        }
        return res;
    }

    void swap(int[] array, int first, int second) {
        int temp = array[first];
        array[first] = array[second];
        array[second] = temp;
    }
}